### PR TITLE
feat(checkbox): add indeterminate checkbox support

### PR DESCRIPTION
**Pull request by DerekLouie**
@gmoothart @ErinCoughlan @jelbourn @ThomasBurleson 

This adds the indeterminate state to md-checkbox.

Please Review. 

References issue #1912

indetcheckbox
indetunselectall
indetselectall

### CODE REVIEW

**Comment by ThomasBurleson** [comment]
@DerekLouie - please note that we have a [PR and issue naming convention](https://github.com/angular/material/blob/master/CONTRIBUTING.md#-commit-message-format).

e.g. (for this PR)  `feat(checkbox): add indeterminate checkbox support`

File: src/components/checkbox/checkbox.js [hunk 1]
```diff
@@ -171,12 +197,48 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
...
+
         if(ngModelCtrl.$viewValue) {
           element.addClass(CHECKED_CSS);
         } else {
           element.removeClass(CHECKED_CSS);
         }
       }
+
+      function setIndeterminateState() {
+        var isIndeterminate = element.attr('indeterminate') === 'true' || false;
```
Comment (DevVersion) [inline src/components/checkbox/checkbox.js]: @DerekLouie I think it would be better, if you use 
```js
$mdUtils.parseAttributeValue(attr.indeterminate)
```

File: src/components/checkbox/checkbox.js [hunk 2]
```diff
@@ -171,12 +197,48 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
...
+          } else {
+            element.removeClass(CHECKED_CSS);
+          }
+        }
+
+        if (isIndeterminate) {
+          element[0].indeterminate = true;
+        } else {
+          element[0].indeterminate = undefined;
+        }
```
Comment (DevVersion) [inline src/components/checkbox/checkbox.js]: I think a ternary operator will be better here:

```js
element[0].indeterminate = isIndeterminate ? true : undefined;
```

File: src/components/checkbox/checkbox.scss [hunk 3]
```diff
@@ -105,6 +105,24 @@ $checkbox-top: 12px !default;
     cursor: default;
   }
 
+  &[indeterminate="true"] ._md-icon{
```
Comment (DevVersion) [inline src/components/checkbox/checkbox.scss]: Can you not apply a custom class to the element (like `.md-indeterminate`)? 

This will be more compatible with the attribute value.

File: src/components/checkbox/checkbox.js [hunk 4]
```diff
@@ -69,6 +80,7 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
 
   function compile (tElement, tAttrs) {
     var container = tElement.children();
+    var indeterminateStateEnabled = tElement[0].hasAttribute('indeterminate');
```
Comment (crisbeto) [inline src/components/checkbox/checkbox.js]: You can replace this one with `tAttrs.hasOwnProperty('indeterminate')` so you don't have to touch the DOM.

File: src/components/checkbox/demoSelectAll/script.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    }
+  };
+
+  $scope.isChecked = function() {
+    if ($scope.selected.length === $scope.items.length) {
+      return true;
+    } else {
+      return false;
+    }
+  };
```
Comment (crisbeto) [inline src/components/checkbox/demoSelectAll/script.js]: This whole check can be replaced with `return $scope.selected.length === $scope.items.length`.

File: src/components/checkbox/checkbox.js [hunk 6]
```diff
@@ -156,11 +175,18 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
...
       }
+
       function listener(ev) {
         if (element[0].hasAttribute('disabled')) {
           return;
         }
 
+        if (indeterminateStateEnabled) {
+          if (attr.indeterminateClick) {
+            scope.$apply(element.attr('indeterminate-click'));
```
Comment (crisbeto) [inline src/components/checkbox/checkbox.js]: You should be able to rewrite this one as `scope.$apply(attr.indeterminateClick);`.

File: src/components/checkbox/checkbox.js [hunk 7]
```diff
@@ -156,11 +175,18 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
           listener(ev);
         }
       }
+
       function listener(ev) {
         if (element[0].hasAttribute('disabled')) {
```
Comment (crisbeto) [inline src/components/checkbox/checkbox.js]: This check can be rewritten as `attr.hasOwnProperty('disabled')`.

File: src/components/checkbox/checkbox.js [hunk 8]
```diff
@@ -171,12 +197,48 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
...
+        // mutually exclusive states.
+        if (attr.indeterminateCheckedWhen) {
+          var isChecked = scope.$eval(attr.indeterminateCheckedWhen);
+          if (isChecked) {
+            element.addClass(CHECKED_CSS);
+            isIndeterminate = false;
+          } else {
+            element.removeClass(CHECKED_CSS);
+          }
+        }
```
Comment (crisbeto) [inline src/components/checkbox/checkbox.js]: This whole block could be shortened down to:

```javascript
var isChecked = scope.$eval(attr.indeterminateCheckedWhen);

element.toggleClass(CHECKED_CSS, isChecked);
isIndeterminate = !isChecked;
```

File: src/components/checkbox/checkbox.js (hunk 6 above)
Comment (DevVersion) [inline src/components/checkbox/checkbox.js]: Also I think, that using `$apply` for that is not the right way.

Why aren't you using `$parse(attr.indeterminateClick)(scope)`?

File: src/components/checkbox/demoSelectAll/script.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    return list.indexOf(item) > -1;
+  };
+
+  $scope.isIndeterminate = function() {
+    if ($scope.selected.length !== 0 &&
+        $scope.selected.length !== $scope.items.length) {
+      return true;
+    } else {
+      return false;
+    }
```
Comment (crisbeto) [inline src/components/checkbox/demoSelectAll/script.js]: This check can be replaced with `return $scope.selected.length !== 0 && $scope.selected.length !== $scope.items.length;`.

File: src/components/checkbox/demoSelectAll/script.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
+
+angular.module('checkboxDemo3', ['ngMaterial'])
+
+.controller('AppCtrl', function($scope) {
+  $scope.items = [1,2,3,4,5];
+  $scope.selected = [1];
+  $scope.toggle = function (item, list) {
+    var idx = list.indexOf(item);
+    if (idx > -1) list.splice(idx, 1);
+    else list.push(item);
```
Comment (crisbeto) [inline src/components/checkbox/demoSelectAll/script.js]: Not sure whether writing an `if/else` like this matches the guidelines.

File: src/components/checkbox/checkbox.js (hunk 6 above)
Comment (crisbeto) [inline src/components/checkbox/checkbox.js]: Isn't that the same as `scope.$eval(attr.indeterminateClick)`?

Comment (DevVersion) [inline src/components/checkbox/checkbox.js]: :+1: Oh I think an `$evalAsync` is better here, because it's evaling in an listener. Thoughts?

Comment (crisbeto) [inline src/components/checkbox/checkbox.js]: Looks like it. Can't be sure since I haven't used `$evalAsync` myself.

File: src/components/checkbox/demoSelectAll/script.js (hunk 10 above)
Comment (DerekLouie) [inline src/components/checkbox/demoSelectAll/script.js]: Yeah I noticed that (this style got Apple in a lot of trouble not too long ago..), this code is pasted from [this file](https://github.com/angular/material/blob/master/src/components/checkbox/demoSyncing/script.js)

Should I can change it in both places?

File: src/components/checkbox/checkbox.js (hunk 8 above)
Comment (DerekLouie) [inline src/components/checkbox/checkbox.js]: I do not believe this is logically equivalent

Comment (crisbeto) [inline src/components/checkbox/checkbox.js]: Fair enough, I'm probably missing a bit of context.

File: src/components/checkbox/checkbox.js (hunk 6 above)
Comment (DerekLouie) [inline src/components/checkbox/checkbox.js]: Yes, eval would be the correct thing to use here, and I should have noticed the `if` statement had the `attr` I needed :+1:.

As a sidenote: I tried substituting `$evalAsync` in here but it breaks necessary indet state logic.

This is because if this runs asynchronously it does not skip the code I need it to in the `render` function which ensures when a checkbox is indeterminate it does not adhere to the normal "checked' / 'unchecked' click logic.

Comment (topherfangio) [inline src/components/checkbox/checkbox.js]: I ran into some issues with the Speed Dial when I was using `$apply` instead of `$evalAsync` because there were cases when the code could be called during an existing digest and throw an error.

If I recall, this occurred when a user added a custom `ng-click` to the component, so you may want to double-check that case.

File: src/components/checkbox/checkbox.js (hunk 1 above)
Comment (DerekLouie) [inline src/components/checkbox/checkbox.js]: Ahh so the issue with this is that `$mdUtil.parseAttributeBoolean` will interpret an empty string as `true` which we want instead to be interpreted as `false` (in the case no value is passed in).

Comment (DevVersion) [inline src/components/checkbox/checkbox.js]: @DerekLouie Yes, that's by intention. Shouldn't `<md-checkbox inderterminate>` work?

Comment (topherfangio) [inline src/components/checkbox/checkbox.js]: The convention we usually follow is that if the attribute is found with no value, then it is treated as `true` as this follows standard HTML attributes (readonly, disabled, etc).

So,

 - `intermediate` == true
 - `intermediate="true"` == true
 - `intermediate="false"` == false

@EladBezalel @ThomasBurleson Please correct me if I'm wrong here.

Comment (DevVersion) [inline src/components/checkbox/checkbox.js]: @topherfangio I agree with you. That's exactly what `$mdUtil.parseAttributeValue` does.

But notice, attributes like `readonly`, `disabled` or `required` should ignore false values.
That's why you should use in that case
```js
$mdUtil.parseAttributeValue(attr, false)
```

**Comment by DerekLouie** [comment]
Hello All,

I did a minor refactor. 

The appearance of the indeterminate state is now class based instead of attribute selector based. This allowed me to get rid of the "indeterminate-when" attribute and simplified the api a little.

I also fixed the aforementioned comments.

Please Review.

Best,

Derek

File: src/components/checkbox/checkbox.spec.js [hunk 11]
```diff
@@ -246,5 +247,115 @@ describe('mdCheckbox', function() {
       expect(isChecked(checkbox)).toBe(false);
       expect(checkbox.hasClass('ng-invalid')).toBe(true);
     });
+
+    describe('with the indeterminate attribute', function() {
+
+      it('should set indeterminate attr to false by default', function() {
```
Comment (topherfangio) [inline src/components/checkbox/checkbox.spec.js]: I disagree with this test; I believe it should be set to true by default to follow our existing conventions.

File: src/components/checkbox/checkbox.js (hunk 1 above)
Comment (DerekLouie) [inline src/components/checkbox/checkbox.js]: Ahh so it does seem like `parseAttributeValue` does not exist. BUT `parseAttributeBoolean` does exist. Unfortunately because we can accept a function now as well as a string, we can't use `parseAttributeBoolean` :(

File: src/components/checkbox/checkbox.spec.js (hunk 11 above)
Comment (DerekLouie) [inline src/components/checkbox/checkbox.spec.js]: Hmm.. that does make a lot of sense :).

Changed, thank you.

File: src/components/checkbox/checkbox.js (hunk 6 above)
Comment (DerekLouie) [inline src/components/checkbox/checkbox.js]: Added an ng-click to the demo locally, both functions fired correctly, so we are all good on this front!

File: src/components/checkbox/checkbox.js [hunk 12]
```diff
@@ -27,6 +27,16 @@ angular
...
  * @param {string=} aria-label Adds label to checkbox for accessibility.
+ * @param {expression=} indeterminate This determines when the checkbox
+ * should be rendered as 'indeterminate'.
+ * If a truthy expression or no value is passed in the checkbox renders in the indeterminate state.
+ * If falsy expression is passed in it just looks like a normal unchecked checkbox.
+ * To use the other indeterminate-* attributes this attribute must be present.
+ * @param {expression=} indeterminate-checked-when This determines when the
+ * indeterminate checkbox should appear checked. The 'indeterminate' state is mutually
+ * exclusive with the 'checked' and 'unchecked' states.
+ * @param {expression=} indeterminate-click This gets triggered when the
```
Comment (ThomasBurleson) [inline src/components/checkbox/checkbox.js]: I do not understand the feature here. At first glance, this seems to interfere with ng-click. Or is this an expression that should be invoked with the indeterminant state changes?

e.g.  `md-change-indeterminant`

What is the use case here ?

File: src/components/checkbox/checkbox.js [hunk 13]
```diff
@@ -27,6 +27,16 @@ angular
  * @param {string=} ng-change Angular expression to be executed when input changes due to user interaction with the input element.
  * @param {boolean=} md-no-ink Use of attribute indicates use of ripple ink effects
  * @param {string=} aria-label Adds label to checkbox for accessibility.
+ * @param {expression=} indeterminate This determines when the checkbox
+ * should be rendered as 'indeterminate'.
+ * If a truthy expression or no value is passed in the checkbox renders in the indeterminate state.
+ * If falsy expression is passed in it just looks like a normal unchecked checkbox.
+ * To use the other indeterminate-* attributes this attribute must be present.
+ * @param {expression=} indeterminate-checked-when This determines when the
```
Comment (ThomasBurleson) [inline src/components/checkbox/checkbox.js]: This appears to be a confusing API. Perhaps:

`md-indeterminant="<expression>"`

when the `expression` becomes true forces the checkbox to indeterminant state.

File: src/components/checkbox/checkbox.js (hunk 12 above)
Comment (DerekLouie) [inline src/components/checkbox/checkbox.js]: You are correct, this is not a good attribute.

Previously when I was unsure if a check box could be selected AND indeterminate I thought the click function would be useful for special toggle cases, but in retrospect this was a poor api design decision. Removing it now.

File: src/components/checkbox/checkbox.js (hunk 13 above)
Comment (DerekLouie) [inline src/components/checkbox/checkbox.js]: I have made some naming changes. Not completely sure if this is what you meant?

Please give the new code a look-over, any notes on how to give the api more clarity are appreciated!

File: src/components/checkbox/checkbox.spec.js [hunk 14]
```diff
@@ -246,5 +247,96 @@ describe('mdCheckbox', function() {
...
+        var element = compileAndLink(
+            '<div>' +
+            '<md-checkbox md-indeterminant></md-checkbox>' +
+            '</div>');
+
+        var checkbox = element.find('md-checkbox');
+
+        pageScope.$apply();
+
+        expect(checkbox.hasClass(INDETERMINATE_CSS)).toBe(true);
```
Comment (ErinCoughlan) [inline src/components/checkbox/checkbox.spec.js]: Is `toHaveClass` copied over from our branch? This would be a lot easier to read as:

`expect(checkbox).toHaveClass(INDETERMINATE_CSS);`

File: src/components/checkbox/checkbox.spec.js [hunk 15]
```diff
@@ -246,5 +247,96 @@ describe('mdCheckbox', function() {
       expect(isChecked(checkbox)).toBe(false);
       expect(checkbox.hasClass('ng-invalid')).toBe(true);
     });
+
+    describe('with the md-indeterminant attribute', function() {
+
+      it('should set md-indeterminant attr to true by default', function() {
+        var element = compileAndLink(
```
Comment (ErinCoughlan) [inline src/components/checkbox/checkbox.spec.js]: Do you need the wrapper `div`? Would `var checkbox = compileAndLink('<md-checkbox ... </md-checkbox>');` work?

File: src/components/checkbox/checkbox.spec.js [hunk 16]
```diff
@@ -246,5 +247,96 @@ describe('mdCheckbox', function() {
...
+            '<div>' +
+            '<md-checkbox md-indeterminant></md-checkbox>' +
+            '</div>');
+
+        var checkbox = element.find('md-checkbox');
+
+        pageScope.$apply();
```
Comment (ErinCoughlan) [inline src/components/checkbox/checkbox.spec.js]: `compileAndLink` calls `pageScope.$apply();`. Do you need another one?

File: src/components/checkbox/demoBasicUsage/index.html [hunk 17]
```diff
@@ -43,6 +43,18 @@
...
         </div>
+        <div flex-xs flex="50">
+          <md-checkbox md-indeterminant
+            aria-label="Checkbox Indeterminate" class="md-primary">
+            Checkbox: Indeterminate
+          </md-checkbox>
+        </div>
+        <div flex-xs flex="50">
+          <md-checkbox md-indeterminant aria-label="Checkbox Disabled Indeterminate"
+            ng-disabled="true" class="md-primary">
```
Comment (ErinCoughlan) [inline src/components/checkbox/demoBasicUsage/index.html]: nit: indent to line up with md-indeterminate or indent 4 spaces instead of 2 to separate attributes from the checkbox content. (Ditto above)

**Comment by DerekLouie** [comment]
Hello All!

Did a significant re-write (thank you @ErinCoughlan for the help). 

Please check it out!

Best,

Derek

File: src/components/checkbox/checkbox.js [hunk 18]
```diff
@@ -27,7 +27,12 @@ angular
  * @param {string=} ng-change Angular expression to be executed when input changes due to user interaction with the input element.
  * @param {boolean=} md-no-ink Use of attribute indicates use of ripple ink effects
  * @param {string=} aria-label Adds label to checkbox for accessibility.
- * Defaults to checkbox's text. If no default text is found, a warning will be logged.
+ *     Defaults to checkbox's text. If no default text is found, a warning will be logged.
+ * @param {expression=} md-indeterminant This determines when the checkbox
```
Comment (jelbourn) [inline src/components/checkbox/checkbox.js]: Spelling: `indeterminate` here and everywhere else

File: src/components/checkbox/checkbox.js [hunk 19]
```diff
@@ -27,7 +27,12 @@ angular
...
+ *     should be rendered as 'indeterminant'.
+ *     If a truthy expression or no value is passed in the checkbox renders in the md-indeterminant state.
+ *     If falsy expression is passed in it just looks like a normal unchecked checkbox.
+ *     To use the other md-indeterminant-* attributes this attribute must be present.
```
Comment (jelbourn) [inline src/components/checkbox/checkbox.js]: Documentation needs to explain the interaction between `indeterminate` and `checked`

File: src/components/checkbox/checkbox.js [hunk 20]
```diff
@@ -89,8 +95,14 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
     });
 
     return function postLink(scope, element, attr, ngModelCtrl) {
+      var isIndeterminant;
       ngModelCtrl = ngModelCtrl || $mdUtil.fakeNgModel();
       $mdTheming(element);
+      if (mdIndeterminantStateEnabled) {
+        // Bootstrap initial state
```
Comment (jelbourn) [inline src/components/checkbox/checkbox.js]: This comment doesn't say anything useful.

File: src/components/checkbox/checkbox.js [hunk 21]
```diff
@@ -171,12 +184,18 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
...
-        if(ngModelCtrl.$viewValue) {
+        if(ngModelCtrl.$viewValue && !isIndeterminant) {
           element.addClass(CHECKED_CSS);
         } else {
           element.removeClass(CHECKED_CSS);
         }
       }
+
+      function setIndeterminantState() {
+        isIndeterminant = scope.$eval(attr.mdIndeterminant) === false ? false : true;
```
Comment (jelbourn) [inline src/components/checkbox/checkbox.js]: `$scope.$watch` evaluates the expression already, so running `$eval` again is doing extra, unnecessary work.

Comment (jelbourn) [inline src/components/checkbox/checkbox.js]: This assignment is kind of clunky. I'd do
```js
isIndeterminant = indeterminateValue || indeterminateValue === '';
```

File: src/components/checkbox/checkbox.js [hunk 22]
```diff
@@ -171,12 +184,18 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
...
+        element[0].indeterminate = isIndeterminant ? true : undefined;
```
Comment (jelbourn) [inline src/components/checkbox/checkbox.js]: Why is setting a property to the element necessary, and why does it need to be either `true` or `undefined`?

File: src/components/checkbox/checkbox.spec.js [hunk 23]
```diff
@@ -246,5 +247,45 @@ describe('mdCheckbox', function() {
...
+
+    describe('with the md-indeterminant attribute', function() {
+
+      it('should set md-indeterminant attr to true by default', function() {
+        var checkbox = compileAndLink('<md-checkbox md-indeterminant></md-checkbox>');
+
+        expect(checkbox).toHaveClass(INDETERMINATE_CSS);
+      });
+
+      it('should set its HTMLInputElement.indeterminate to be undefined when not md-indeterminant', function() {
```
Comment (DevVersion) [inline src/components/checkbox/checkbox.spec.js]: I would rather say: `should not set the indeterminate property when md-indeterminate is false.`

File: src/components/checkbox/demoSelectAll/style.css [hunk 24]
```diff
@@ -0,0 +1,13 @@
+legend {
```
Comment (jelbourn) [inline src/components/checkbox/demoSelectAll/style.css]: Prefix demo styles with `demo`-

File: src/core/style/mixins.scss [hunk 25]
```diff
@@ -251,4 +251,21 @@
     cursor: default;
   }
 
+  &.md-indeterminant ._md-icon{
+    &:after {
+        box-sizing: border-box;
```
Comment (jelbourn) [inline src/core/style/mixins.scss]: Too much indent. You're IDE is likely configured to the wrong indent for `.scss` files.

File: src/components/checkbox/checkbox.js (hunk 19 above)
Comment (DevVersion) [inline src/components/checkbox/checkbox.js]: In addition to that, I thought `md-indeterminat-*` attribute have been dropped? So maybe remove that line.

File: src/components/checkbox/checkbox.spec.js [hunk 26]
```diff
@@ -246,5 +247,45 @@ describe('mdCheckbox', function() {
...
+
+      it('should set its HTMLInputElement.indeterminate to be "true" when md-indeterminant', function() {
+        var checkbox = compileAndLink('<md-checkbox md-indeterminant></md-checkbox>');
+
+        expect(checkbox[0].indeterminate).toBe(true);
+      });
+
+      it('should be set "md-indeterminant" class according to a passed in function', function() {
+        pageScope.isIndeterminate = function() { return true; } ;
+
```
Comment (DevVersion) [inline src/components/checkbox/checkbox.spec.js]: @DerekLouie Maybe you can also add a test, which changes the `indeterminate` expression on runtime as well?

File: src/components/checkbox/checkbox.js (hunk 22 above)
Comment (DerekLouie) [inline src/components/checkbox/checkbox.js]: I believe this is to leverage some native browser implementation features. @ErinCoughlan?

File: src/components/checkbox/checkbox.js (hunk 21 above)
Comment (DerekLouie) [inline src/components/checkbox/checkbox.js]: Hmm so, the $scope.$watch will pass in the new value as the first parameter which will either be - `undefined`, `true`, or `false`.

If it returns `false` we want to set `isIndeterminate` to `false`. 

All other cases we want to set  `isIndeterminate` to `true`.

I cannot seem to find a better way to achieve that, but I am open to suggestions!

File: src/core/style/mixins.scss [hunk 27]
```diff
@@ -251,4 +251,21 @@
     cursor: default;
   }
 
+  &.md-indeterminate ._md-icon{
```
Comment (EladBezalel) [inline src/core/style/mixins.scss]: Why is that in here and not in `checkbox.scss`?

File: src/core/style/themes.scss [hunk 28]
```diff
@@ -39,4 +39,10 @@ html, body {
   &#{$checkedSelector} ._md-icon:after {
     border-color: '{{primary-contrast-0.87}}';
   }
+
+  & .md-indeterminate[disabled] {
```
Comment (EladBezalel) [inline src/core/style/themes.scss]: Why is that in here and not in `checkbox.scss`?

File: src/core/style/mixins.scss (hunk 27 above)
Comment (ErinCoughlan) [inline src/core/style/mixins.scss]: The checkbox CSS is used both within the checkbox and the select menu (for multi-select case).

File: src/core/style/themes.scss (hunk 28 above)
Comment (DerekLouie) [inline src/core/style/themes.scss]: We created a mixin for the checkbox's css so it could be used in other components. This is where all the css goes now that might be used by multiple components.

**Comment by DerekLouie** [comment]
Hello,

I have addressed the aforementioned comments, please review!

Best,

Derek

File: test/angular-material-spec.js [hunk 29]
```diff
@@ -194,10 +200,71 @@
...
+
+      /**
+       * Asserts that a given element contains a given substring in
+       * its innerHTML property.
+       * Accepts any of:
+       *   {string} - A CSS selector.
+       *   {angular.JQLite} - The result of a jQuery query.
+       *   {Element} - A DOM element.
+       */
+      toContainHtml: function() {
```
Comment (DevVersion) [inline test/angular-material-spec.js]: Why has that been added?

Comment (ErinCoughlan) [inline test/angular-material-spec.js]: I'm okay with waiting to add this method until we merge feat/panel. It was added here for completeness and convenience, but it's not being used yet.

**Comment by ErinCoughlan** [comment]
LGTM.

Make sure to change your commit message to be:
feat(checkbox): add indeterminate checkbox support

**Comment by DerekLouie** [comment]
Changed the commit message, please review again at your earliest convenience!

File: test/angular-material-spec.js (hunk 29 above)
Comment (DerekLouie) [inline test/angular-material-spec.js]: I believe I these methods will be useful, and am adding all of them now (instead of just the `toHaveClass` method changes) so later down the line we don't get merge conflicts when we merge feat/panel.

File: src/components/checkbox/checkbox.js [hunk 30]
```diff
@@ -171,12 +183,18 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
...
-        if(ngModelCtrl.$viewValue) {
+        if(ngModelCtrl.$viewValue && !isIndeterminate) {
           element.addClass(CHECKED_CSS);
         } else {
           element.removeClass(CHECKED_CSS);
         }
       }
+
+      function setIndeterminateState(newValue) {
+        isIndeterminate = (newValue === false) ? false : true;
```
Comment (jelbourn) [inline src/components/checkbox/checkbox.js]: isIndeterminate = newValue !== false;

File: test/angular-material-spec.js (hunk 29 above)
Comment (DevVersion) [inline test/angular-material-spec.js]: Didn't see actual point, because we rarely change the spec shims. So there won't be any merge conflicts. But it's okay.

File: src/components/checkbox/checkbox.js [hunk 31]
```diff
@@ -171,12 +183,18 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
...
+        element[0].indeterminate = isIndeterminate ? true : undefined;
```
Comment (jelbourn) [inline src/components/checkbox/checkbox.js]: It's still not clear why you need to set an `indeterminate` property on the DOM element.

File: src/components/checkbox/checkbox.spec.js [hunk 32]
```diff
@@ -246,5 +247,64 @@ describe('mdCheckbox', function() {
...
+      it('should set md-indeterminate attr to true by default', function() {
+        var checkbox = compileAndLink('<md-checkbox md-indeterminate></md-checkbox>');
+
+        expect(checkbox).toHaveClass(INDETERMINATE_CSS);
+      });
+
+      it('should not set the HTMLInputElement.indeterminate property when md-indeterminate is false', function() {
+        var checkbox = compileAndLink('<md-checkbox md-indeterminate="false"></md-checkbox>');
+
+        expect(checkbox[0].indeterminate).toBe(undefined);
```
Comment (jelbourn) [inline src/components/checkbox/checkbox.spec.js]: `toBeUndefined()`

File: src/components/checkbox/demoSelectAll/style.css [hunk 33]
```diff
@@ -0,0 +1,15 @@
...
+  fieldset.standard {
+    border-style: solid;
+    border-width: 1px;
+    height: 100%;
+  }
+
+  .select-all-checkboxes {
+    padding-left: 30px;
+  }
+}
```
Comment (jelbourn) [inline src/components/checkbox/demoSelectAll/style.css]: Just do these as
```css
.demo-legend { ... }
.demo-fieldset { ... }
.demo-select-all-checkboxes { .. }
```

File: test/angular-material-spec.js (hunk 29 above)
Comment (jelbourn) [inline test/angular-material-spec.js]: It's better to not add things until they're used. Plus you'll prevent merge conflicts.

Comment (DevVersion) [inline test/angular-material-spec.js]: Yep, I agree. These test helpers, should be just merged within the `feat/panel` branch. Partially merging code snippets doesn't make any sense.

**Comment by ErinCoughlan** [comment]
@jelbourn Should there be aria-checked somewhere in here? I didn't see it in the initial code, so is it not helpful?

File: test/angular-material-spec.js (hunk 29 above)
Comment (DerekLouie) [inline test/angular-material-spec.js]: Fair point, taking out the non-used additions!

File: src/components/checkbox/checkbox.js (hunk 31 above)
Comment (DerekLouie) [inline src/components/checkbox/checkbox.js]: Apologies, I miss-understood the spec I was looking at. This is for when we move to the native checkbox. The only benefit of this would be that users can use the `:indeterminate` pseudo selector to find the indeterminate `md-checkbox`es.

@jelbourn I am okay with removing this code, do you care whether or not the directive is query-able using the  `:indeterminate` pseudo selector?

**Comment by DerekLouie** [comment]
Took out some extra code and also added some code to dictate the 'aria-checked' attribute.

Please review.

File: src/components/checkbox/checkbox.js [hunk 34]
```diff
@@ -171,12 +183,20 @@ function MdCheckboxDirective(inputDirective, $mdAria, $mdConstant, $mdTheming, $
...
           element.addClass(CHECKED_CSS);
         } else {
           element.removeClass(CHECKED_CSS);
         }
       }
+
+      function setIndeterminateState(newValue) {
+        isIndeterminate = newValue !== false;
+        if (isIndeterminate) {
+          element.attr('aria-checked', 'mixed');
```
Comment (ErinCoughlan) [inline src/components/checkbox/checkbox.js]: aria-checked is never set for checked or unchecked checkboxes either. If you're going to set it to "mixed", you also should set it to true/false where appropriate.

Comment (DerekLouie) [inline src/components/checkbox/checkbox.js]: It actually is set, and accounted for in the [tests](https://github.com/DerekLouie/material/blob/inteterminateCheckbox/src/components/checkbox/checkbox.spec.js#L43). It was a little curious though that the code isn't found in this file..

Comment (jelbourn) [inline src/components/checkbox/checkbox.js]: I believe it's already handled by ngAria

**Comment by DerekLouie** [comment]
Hello All,

I have addressed all the aforementioned comments, please review at your earliest convenience!

Best,

Derek

File: src/core/style/mixins.scss (hunk 27 above)
Comment (DevVersion) [inline src/core/style/mixins.scss]: @DerekLouie: Nitpicky -  Please space bracket from `._md-icon`

**Comment by DevVersion** [comment]
LGTM now, aside one comment.

**Comment by DerekLouie** [comment]
Apologies, there have been quite a few comments, I will happily address it, which comment?

**Comment by DerekLouie** [comment]
Ahh thank you @DevVersion, fixed the comment.

**Comment by jelbourn** [comment]
LGTM

File: src/core/style/themes.scss (hunk 28 above)
Comment (EladBezalel) [inline src/core/style/themes.scss]: Indeterminate isn't used anywhere else, only on progress and this seems not related, 
Also don't you think this class should be private?

File: src/core/style/mixins.scss [hunk 35]
```diff
@@ -251,4 +251,21 @@
     cursor: default;
   }
 
+  &.md-indeterminate ._md-icon {
```
Comment (EladBezalel) [inline src/core/style/mixins.scss]: This is not related here as well, this is very specific for checkbox

File: src/core/style/themes.scss (hunk 28 above)
Comment (topherfangio) [inline src/core/style/themes.scss]: @EladBezalel In response to private: I had a discussion with Thomas where we thought that CSS classes representing the state of a component should probably be public as they are likely the ones to be used most.

**Comment by crisbeto** [comment]
LGTM

**Comment by ThomasBurleson** [comment]
@DerekLouie - really nice feature. I LOVE this demo:

screen shot 2016-03-30 at 12 27 27 pm

File: src/core/style/mixins.scss (hunk 35 above)
Comment (DerekLouie) [inline src/core/style/mixins.scss]: @EladBezalel I could remove the `&`. The rule would still be valid and it would be one class less specific?
